quickjs.c

```diff
@@ -47302,6 +47314,52 @@ uint8_t *JS_GetArrayBuffer(JSContext *ctx, size_t *psize, JSValueConst obj)
     return NULL;
 }
 
+// ES #sec-arraybuffer.prototype.transfer
+static JSValue js_array_buffer_transfer(JSContext *ctx,
+                                        JSValueConst this_val,
+                                        int argc, JSValueConst *argv)
+{
+    JSArrayBuffer *abuf;
+    uint64_t new_len, old_len;
+    uint8_t *bs, *new_bs;
+
+    abuf = JS_GetOpaque2(ctx, this_val, JS_CLASS_ARRAY_BUFFER);
+    if (!abuf)
+        return JS_EXCEPTION;
+    if (abuf->shared)
+        return JS_ThrowTypeError(ctx, "cannot transfer a SharedArrayBuffer");
+    if (argc < 1 || JS_IsUndefined(argv[0]))
+        new_len = abuf->byte_length;
+    else if (JS_ToIndex(ctx, &new_len, argv[0]))
+        return JS_EXCEPTION;
+    if (abuf->detached)
+        return JS_ThrowTypeErrorDetachedArrayBuffer(ctx);
+    /* create an empty AB */
+    if (new_len == 0) {
+        JS_DetachArrayBuffer(ctx, this_val);
+        return js_array_buffer_constructor2(ctx, JS_UNDEFINED, 0, JS_CLASS_ARRAY_BUFFER);
+    }
+    bs = abuf->data;
+    old_len = abuf->byte_length;
+    /* neuter the backing buffer */
+    abuf->data = NULL;
+    abuf->byte_length = 0;
+    abuf->detached = TRUE;
+    /* if length mismatch, realloc. Otherwise, use the same backing buffer. */
+    if (new_len != old_len) {
+        new_bs = js_realloc(ctx, bs, new_len);
+        if (!new_bs)
+            return JS_EXCEPTION;
```

Comment:
Leaks memory:
```suggestion
        if (!new_bs) {
            js_free(ctx, bs);
            return JS_EXCEPTION;
        }
```
Either ^ or the original arraybuffer should stay untouched. The second option is probably least surprising to users.